Initial commit of hof-frontend-assets

JoeChapman opened the pull request:
This will replace the assets folder currently copied to each service

- add assets
- add execultable to build assets to public in consumer
- add meta stuff (gitignore ...)
- add readme doc

JoeChapman committed:
Initial commit of hof-frontend-assets

- add assets
- add execultable to build assets to public in consumer
- add meta stuff (gitignore ...)
- add readme doc

easternbloc commented on js/index.js:
```diff
@@ -0,0 +1,75 @@
+'use strict';
```
this should be tested when I did it for BRP it was rushed out. I think we should do it properly now

JoeChapman replied:
Presumably you don't mean `use strict` should be tested, you mean the js?

JoeChapman replied:
if so, fair comment

easternbloc replied:
no I mean `use strict`

easternbloc replied:
of course I mean the FILE!

JoeChapman committed:
Refactor code and test

- refactor typeahead, arrange in to "classes"
- test frontend code with karma
- separate tests and coverage for browser and unit
- add meta files, ignore, travis snyk

joefitter commented on README.md:
```diff
@@ -1,2 +1,42 @@
+To use Typeahead Aria, add `class='typeahead' to the HTML `select` element you wish to decorate.
+
+### Country List
+Add [Typeahead Countries](https://github.com/UKHomeOffice/typeahead-countries) to the field in your HOF service that requires the list.
```
Isn't this a separate module - homeoffice-countries?

JoeChapman replied:
mistake

joefitter commented on README.md:
```diff
@@ -1,2 +1,42 @@
+    className: 'visuallyhidden'
+  },
+  validate: ['required']
+},
```
These docs are repeated, homeoffice-countries is a thing, or do we not want to use that yet?

JoeChapman replied:
mistake

joefitter reviewed (commented):
ideally we'll lose all this custom sass at some point, obviously better than what we currently have though!

JoeChapman committed:
Address comments for typeahead countries and fix linting

easternbloc commented on README.md:
```diff
@@ -1,2 +1,40 @@
+  "hof-frontend-assets": "^0.1.0"
+},
+"scripts": {
+  "postinstall": "npm hof-frontend-assets"
```
delete `npm`

easternbloc commented on README.md:
````diff
@@ -1,2 +1,40 @@
+
+
+```
+"dependencies": {
````
Remove this. We should just talk about using a typeahead here.

easternbloc commented on README.md:
````diff
@@ -1,2 +1,40 @@
+}
+```
+
+### Country List
````
Remove this.

easternbloc commented on bin/tasks.js:
```diff
@@ -0,0 +1,14 @@
+// jscs:disable maximumLineLength
+module.exports = {
+  'make-folders': `mkdir -p ${target}/js ${target}/css ${target}/images`,
+  'compile-css': `${cwd}/node_modules/.bin/node-sass ${source}/scss/app.scss ${target}/css/app.css --include-path ./node_modules`,
```
could maybe use https://www.npmjs.com/package/witch to make this read nicer...

easternbloc commented on package.json:
```diff
@@ -0,0 +1,69 @@
+  "version": "0.1.0",
+  "main": "./bin/index.js",
+  "description": "Frontend JS, CSS and images for a form service",
+  "license": "GPLv3",
```
MIT

easternbloc commented on package.json:
```diff
@@ -0,0 +1,69 @@
+    "cover": "istanbul cover _mocha -- test/unit",
+    "check-coverage": "istanbul check-coverage",
+    "unit:coverage": "npm-run-all cover check-coverage",
+    "snyk": "snyk test",
```
can you run snyk test on pull requests but not master: https://docs.travis-ci.com/user/pull-requests

JoeChapman replied:
Addressed in https://github.com/UKHomeOfficeForms/hof-frontend-assets/pull/2

easternbloc commented on src/js/index.js:
```diff
@@ -0,0 +1,13 @@
+helpers.documentReady(progressiveReveal);
+helpers.documentReady(formFocus);
+
+var typeahead = require('./typeahead');
```
think this should be an exports that you can call with `options`

easternbloc commented on src/js/typeahead/bloodhound.js:
```diff
@@ -0,0 +1,28 @@
+
+Bloodhound.prototype.getData = function getData(list) {
+  return {
+    datumTokenizer: TypeaheadBloodhound.tokenizers.whitespace,
```
too many opinions this should be passed in options

easternbloc commented on src/js/typeahead/helpers.js:
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+module.exports = {
```
this should be a sort option but not the default

easternbloc replied:
or maybe the default but overwritable. The problem here is this is something you'd want to configure based on user research.

JoeChapman committed:
export typeahead

lennym commented on package.json:
```diff
@@ -0,0 +1,69 @@
+    "unit:coverage": "npm-run-all cover check-coverage",
+    "snyk": "snyk test",
+    "snyk:dev": "snyk test --dev",
+    "lint": "eslint .",
```
`eslint` is not listed in the dependencies. Using with `npm@3` it still "works" because it finds the eslint installed as a child of `eslint-config-homeoffice`, but with `npm@2` this is not flattened to the top level, and so fails.

See https://travis-ci.org/UKHomeOfficeForms/hof-frontend-assets/jobs/188810941

JoeChapman replied:
Rather than install eslint independently, I have config'd travis to install npm 3

lennym replied:
That's a really bad solution for something which is trivial to fix by adding the thing we're depending on as a dependency. What possible arguments are there *against* adding eslint as a dependency?

In particular because this is open source software and we're putting arbitrary and undocumented obstacles in place of potential contributors.

JoeChapman replied:
@joefitter can you remind me why we enforce npm 3? I know it's not so we can bypass explicitly adding deps. However, in light of the fact that we are using npm 3, it seemed superfluous to add deps that are already included.

lennym replied:
I don't think it's superfluous to not rely on side-effects for things to work.

JoeChapman committed:
pass options to typeahead and bloodhound

JoeChapman committed:
enforce npm ~3 and node >=4 for travis and user

JoeChapman committed:
use MIT license

JoeChapman committed:
use witch to find node-sass module

JoeChapman committed:
install eslint, do not insist npm 3 in travis

JoeChapman committed:
install jquery, remove underscore

JoeChapman commented:
@easternbloc @lennym @joefitter I have addressed the comments, please can you check this pr again?
